flow/record/base.py

```diff
@@ -338,243 +335,308 @@ class RecordFieldSet(list):
+        "_zip_longest": zip_longest,
+    }
+    for field in all_fields.values():
+        _globals[f"_field_{field.name}"] = field
+
+    exec(code, _globals)
+
+    return _globals[name]
+
 
-    def __init__(self, name, fields=None):
+class RecordDescriptor:
+    name: str = None
+    recordType: type = None
+    _desc_hash: int = None
+    _fields: Mapping[str, RecordField] = None
+    _all_fields: Mapping[str, RecordField] = None
+    _field_tuples: Sequence[Tuple[str, str]] = None
+
+    def __init__(self, name: str, fields: Optional[Sequence[Tuple[str, str]]] = None) -> None:
         if not name:
             raise RecordDescriptorError("Record name is required")
 
+        # Marked for deprecation
         name = to_str(name)
-
         if isinstance(fields, RecordDescriptor):
+            warnings.warn(
+                "RecordDescriptor initialization with another RecordDescriptor is deprecated",
+                DeprecationWarning,
+                stacklevel=2,
+            )
             # Clone fields
             fields = fields.get_field_tuples()
-        elif not fields:
+        elif fields is None:
+            warnings.warn(
+                "RecordDescriptor initialization by string only definition is deprecated",
+                DeprecationWarning,
+                stacklevel=2,
+            )
             name, fields = parse_def(name)
 
-        fields = list([(to_native_str(k), to_str(v)) for k, v in fields])
-
-        contains_keyword = False
-        for fieldtype, fieldname in fields:
-            if not is_valid_field_name(fieldname):
-                raise RecordDescriptorError("Field '{}' is an invalid or reserved field name.".format(fieldname))
-
-            # Reserved Python keywords are allowed as field names, but at a cost.
-            # When a Python keyword is used as a field name, you can't use it as a kwarg anymore
-            # You'll be forced to either use *args or a expanding a dict to kwargs to initialize a record
-            # E.g. Record('from_value', 'and_value') or Record(**{'from': 1, 'and': 2})
-            # You'll also only be able to get or set reserved attributes using getattr or setattr.
-            # Record initialization will also be slower, due to a different (slower) implementation
-            # that is compatible with this method of initializing records.
-            if keyword.iskeyword(fieldname):
-                contains_keyword = True
-
-        self.fields = OrderedDict([(n, RecordField(n, _type)) for _type, n in fields])
-        all_fields = self.get_all_fields()
         self.name = name
+        self._field_tuples = tuple([(to_native_str(k), to_str(v)) for k, v in fields])
+        self.recordType = _generate_record_class(name, self._field_tuples)
+        self.recordType._desc = self
 
-        if not RE_VALID_RECORD_TYPE_NAME.match(name):
-            raise RecordDescriptorError("Invalid record type name")
+    @staticmethod
+    @functools.lru_cache()
```

Comment:
```suggestion
    @functools.cache()
```
I think you can just use `cache` here. Can't this be a constant, by the way?

Reply:
flow.record still targets python 3.7 (by my request), so it still needs to be `lru_cache()`. The functionality is the same.

In my earlier refactor I turned this into the constant `REQUIRED_FIELDS`, but then I had to move things around, because it uses `RecordField()`. So the constant would disappear somewhere between methods, instead of below the currently defined constants such as `RESERVED_FIELDS`.

Also, there is an unit test that tests if you correctly use the required fields (_version always has to come last). It made testing that part a bit harder.

Let me know what you think.

Reply:
Right, forgot that `@cache` is 3.9.

Hm yeah that's annoying. I don't think this is a particularly hot function to perhaps better to just keep it this way.